plugins/inputs/syslog/syslog.go

```diff
@@ -0,0 +1,378 @@
+			acc.AddError(err)
+		}
+	}
+}
+
+func (s *Syslog) listenStream(acc telegraf.Accumulator) {
+	defer s.wg.Done()
+
+	s.connections = map[string]net.Conn{}
+
+	for {
+		conn, err := s.tcpListener.Accept()
+		if err != nil {
+			if !strings.HasSuffix(err.Error(), ": use of closed network connection") {
+				log.Println(err)
+				acc.AddError(err)
+			}
+			break
+		}
+		var tcpConn, _ = conn.(*net.TCPConn)
+		if s.tlsConfig != nil {
+			conn = tls.Server(conn, s.tlsConfig)
+		}
+
+		s.connectionsMu.Lock()
+		if s.MaxConnections > 0 && len(s.connections) >= s.MaxConnections {
+			s.connectionsMu.Unlock()
+			conn.Close()
+			continue
+		}
+		s.connections[conn.RemoteAddr().String()] = conn
+		s.connectionsMu.Unlock()
+
+		if err := s.setKeepAlive(tcpConn); err != nil {
+			acc.AddError(fmt.Errorf("unable to configure keep alive (%s): %s", s.Address, err))
+		}
+
+		go s.handle(conn, acc)
+	}
+
+	s.connectionsMu.Lock()
+	for _, c := range s.connections {
+		c.Close()
+	}
+	s.connectionsMu.Unlock()
+}
+
+func (s *Syslog) removeConnection(c net.Conn) {
+	s.connectionsMu.Lock()
+	delete(s.connections, c.RemoteAddr().String())
+	s.connectionsMu.Unlock()
+}
+
+func (s *Syslog) handle(conn net.Conn, acc telegraf.Accumulator) {
+	defer s.removeConnection(conn)
+	defer conn.Close()
+
+	if s.ReadTimeout != nil && s.ReadTimeout.Duration > 0 {
+		conn.SetReadDeadline(time.Now().Add(s.ReadTimeout.Duration))
+	}
+
+	var p *rfc5425.Parser
+	if s.BestEffort {
+		p = rfc5425.NewParser(conn, rfc5425.WithBestEffort())
+	} else {
+		p = rfc5425.NewParser(conn)
+	}
+
+	p.ParseExecuting(func(r *rfc5425.Result) {
+		s.store(*r, acc)
+	})
+}
+
+func (s *Syslog) setKeepAlive(c *net.TCPConn) error {
+	if s.KeepAlivePeriod == nil {
+		return nil
+	}
+
+	if s.KeepAlivePeriod.Duration == 0 {
+		return c.SetKeepAlive(false)
+	}
+	if err := c.SetKeepAlive(true); err != nil {
+		return err
+	}
+	return c.SetKeepAlivePeriod(s.KeepAlivePeriod.Duration)
+}
+
+func (s *Syslog) store(res rfc5425.Result, acc telegraf.Accumulator) {
+	if res.Error != nil {
+		acc.AddError(res.Error)
+	}
+	if res.MessageError != nil {
+		acc.AddError(res.MessageError)
+	}
+	if res.Message != nil {
+		acc.AddFields("syslog", fields(res.Message), tags(res.Message), s.now())
+	}
+}
+
+func tags(msg *rfc5424.SyslogMessage) map[string]string {
+	ts := map[string]string{}
+	if lvl := msg.SeverityLevel(); lvl != nil {
+		ts["severity"] = strconv.Itoa(int(*msg.Severity()))
+		ts["severity_level"] = *lvl
+	}
+
+	if f := msg.FacilityMessage(); f != nil {
+		ts["facility"] = strconv.Itoa(int(*msg.Facility()))
+		ts["facility_message"] = *f
+	}
+
+	if msg.Hostname() != nil {
+		ts["hostname"] = *msg.Hostname()
+	}
+
+	if msg.Appname() != nil {
+		ts["appname"] = *msg.Appname()
+	}
+
+	return ts
+}
+
+func fields(msg *rfc5424.SyslogMessage) map[string]interface{} {
+	flds := map[string]interface{}{
+		"version": msg.Version(),
+	}
+
+	if msg.Timestamp() != nil {
+		flds["timestamp"] = *msg.Timestamp()
```

Comment:
Should we use the timestamp as the metric timestamp?  What have we decided about dealing with conflicting timestamps with the same series?

Reply:
In a previous version of this plugin we were using it as a metric timestamp.

The reason we have chosen to store it as a fields depends on the fact that the syslog timestamp represents when the event that the message represents has occurred, not when telegraf received it.

Reply:
So at the end as @goller suggested me, assuming system clock is monotonic, we are using its timestamps as metric. Incrementing a time by 1 nanoseconds in the case it is equal to the previous done.

For future reference, in cases where this assumption can not be made or external timestamps are used as metrics, a solution based on bloom filter (ie., definitely not in set versus false-positives) could fit well our needs. I have one in stash.

Anyway should be ok now.